packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,67 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL, SLOT_COMPONENT_TYPEOF_SYMBOL } from './constants';
+import * as React from 'react';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  componentType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+  overrides?: Partial<Props>;
+};
+
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | undefined,
+  options: SlotOptions<Props> & { required: true },
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | null | undefined,
+  options: SlotOptions<Props> & { required?: boolean },
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | null | undefined,
+  options: SlotOptions<Props> & { required?: boolean },
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, overrides, componentType } = options;
+
+  if (shorthand === null || (shorthand === undefined && !required)) {
+    return undefined;
+  }
+  const metadata: SlotComponentMetadata<Props> = { componentType };
```

Comment:
Out of curiosity, why `componentType` instead of `elementType`?

Reply:
Just because 🤷🏼‍♂️. I guess it's because we have `components` property on the `state` and that would be something similar?! but yeah, `elementType` is more concise